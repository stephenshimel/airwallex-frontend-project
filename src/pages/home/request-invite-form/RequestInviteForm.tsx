import React, { useCallback, useMemo } from "react";
import { yupResolver } from "@hookform/resolvers/yup";
import { useForm, SubmitHandler } from "react-hook-form";
import { TFormFields, TRequestInviteForm } from "./types";
import { schema } from "./schema";
import Form from "component/form/Form";
import ErrorBoundary from "utils/ErrorBoundary";
import usePostRequestInviteData from "utils/hooks/usePostRequestInvite";
import { strings } from "constants/strings";
import { TInputProps } from "component/form/types";

const RequestInviteForm = ({ closeModal }: TRequestInviteForm) => {
	const requestInviteFormItems: TInputProps[] = useMemo(
		() => [
			{ name: "name", title: "Full Name" },
			{ name: "email", title: "Email" },
			{ name: "confirmEmail", title: "Confirm Email" },
		],
		[]
	);

	const {
		register,
		handleSubmit,
		formState: { errors },
	} = useForm<TFormFields>({
		resolver: yupResolver(schema),
		reValidateMode: "onSubmit",
	});

	const {
		mutate,
		isLoading,
		isError: isServerError,
		error: serverError,
		isSuccess,
	} = usePostRequestInviteData();

	const onSubmit: SubmitHandler<TFormFields> = useCallback(
		({ name, email }: TFormFields) => {
			mutate({ name, email });
		},
		[mutate]
	);

	return isSuccess ? (
		// submit success page
		<Form
			title={strings.submitSuccessPage.title}
			subTitle={strings.submitSuccessPage.subTitle}
			buttonLabel={strings.submitSuccessPage.buttonLabel}
			formRelated={{ onSubmit: closeModal }}
		/>
	) : (
		// form page
		<ErrorBoundary>
			<Form<TFormFields>
				title={strings.requestInvitePage.requestFormTitle}
				buttonLabel={strings.requestInvitePage.sendButtonLabel}
				formRelated={{
					formItems: requestInviteFormItems,
					isLoading: isLoading,
					isServerError: isServerError,
					serverError: serverError as string,
					onSubmit: handleSubmit(onSubmit),
					errors: errors,
					register: register,
				}}
			/>
		</ErrorBoundary>
	);
};

export default RequestInviteForm;
