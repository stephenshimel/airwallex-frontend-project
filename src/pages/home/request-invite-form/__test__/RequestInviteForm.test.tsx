/* eslint-disable testing-library/render-result-naming-convention */
/* eslint-disable testing-library/no-wait-for-multiple-assertions */
import React from "react";
import renderer from "react-test-renderer";
import { render, fireEvent, waitFor, screen } from "@testing-library/react";
import { QueryClient, QueryClientProvider } from "react-query";
import { rest } from "msw";
import { setupServer } from "msw/node";
import RequestInviteForm from "../RequestInviteForm";

describe("RequestInviteForm", () => {
	const serverPositive = setupServer(
		rest.post(
			"https://l94wc2001h.execute-api.ap-southeast-2.amazonaws.com/prod/fake-auth",
			async (req, res, ctx) => {
				console.log("Mock server handler triggered");

				// parse the request body to get the email
				const { email } = await req.json();

				console.log("email!!!", email);
				// check if the email is "usedemail@airwallex.com"
				if (email === "usedemail@airwallex.com") {
					console.log("400 returned!!!");
					return res(
						ctx.status(400),
						ctx.json({ error: "Email is already in use" })
					);
				}

				return res(
					ctx.delay(200),
					ctx.status(200),
					ctx.json({ message: "success" })
				);
			}
		)
	);

	beforeAll(() => serverPositive.listen());
	afterEach(() => serverPositive.resetHandlers());
	afterAll(() => serverPositive.close());

	const queryClient = new QueryClient();

	// Positive tests:
	it("renders correctly", () => {
		render(
			<QueryClientProvider client={queryClient}>
				<RequestInviteForm />
			</QueryClientProvider>
		);
		expect(screen.getByText("Request an invite")).toBeInTheDocument();
		expect(screen.getByPlaceholderText("Full Name")).toBeInTheDocument();
		expect(screen.getByPlaceholderText("Email")).toBeInTheDocument();
		expect(screen.getByPlaceholderText("Confirm Email")).toBeInTheDocument();
		expect(screen.getByText("Send")).toBeInTheDocument();
	});

	it("initial status matches the snapshot", () => {
		const tree = render(
			<QueryClientProvider client={queryClient}>
				<RequestInviteForm />
			</QueryClientProvider>
		);
		expect(tree).toMatchSnapshot();
	});

	it("shows `Sending, please wait...` if submiting the form correctly, before server response arrives", async () => {
		render(
			<QueryClientProvider client={queryClient}>
				<RequestInviteForm />
			</QueryClientProvider>
		);
		fireEvent.change(screen.getByPlaceholderText("Full Name"), {
			target: { value: "Stephen Shi" },
		});
		fireEvent.change(screen.getByPlaceholderText("Email"), {
			target: { value: "test@test.com" },
		});
		fireEvent.change(screen.getByPlaceholderText("Confirm Email"), {
			target: { value: "test@test.com" },
		});

		fireEvent.click(screen.getByTestId("submit-button"));

		await waitFor(() => {
			expect(screen.getByText("Sending, please wait...")).toBeInTheDocument();
		});
	});

	it("shows the success page and matches snapshot, if the form is submitted, passed all validation rules, and server response is 200", async () => {
		const tree = render(
			<QueryClientProvider client={queryClient}>
				<RequestInviteForm />
			</QueryClientProvider>
		);
		fireEvent.change(screen.getByPlaceholderText("Full Name"), {
			target: { value: "Stephen Shi" },
		});
		fireEvent.change(screen.getByPlaceholderText("Email"), {
			target: { value: "test@test.com" },
		});
		fireEvent.change(screen.getByPlaceholderText("Confirm Email"), {
			target: { value: "test@test.com" },
		});

		fireEvent.click(screen.getByTestId("submit-button"));

		await waitFor(
			() => {
				expect(screen.getByText("All done!")).toBeInTheDocument();
				expect(
					screen.getByText(
						"You will be one of the first to experience Broccoli & Co. when w launch"
					)
				).toBeInTheDocument();
				expect(screen.getByText("OK")).toBeInTheDocument();
			},
			{ timeout: 3000 }
		);

		expect(tree).toMatchSnapshot();
	});

	// Negative tests:

	it("shows error message when form is submitted with empty fields", async () => {
		render(
			<QueryClientProvider client={queryClient}>
				<RequestInviteForm />
			</QueryClientProvider>
		);
		fireEvent.click(screen.getByTestId("submit-button"));
		expect(
			await screen.findByText("Full Name is required")
		).toBeInTheDocument();
	});

	it("shows error message when name is entered but less than 3 letters", async () => {
		render(
			<QueryClientProvider client={queryClient}>
				<RequestInviteForm />
			</QueryClientProvider>
		);
		fireEvent.change(screen.getByPlaceholderText("Full Name"), {
			target: { value: "S" },
		});
		fireEvent.click(screen.getByTestId("submit-button"));
		expect(
			await screen.findByText("Full Name must be at least 3 characters")
		).toBeInTheDocument();
	});

	it("shows error message when email is entered but not a valid email address", async () => {
		render(
			<QueryClientProvider client={queryClient}>
				<RequestInviteForm />
			</QueryClientProvider>
		);
		fireEvent.change(screen.getByPlaceholderText("Email"), {
			target: { value: "wrong-email" },
		});
		fireEvent.click(screen.getByTestId("submit-button"));
		expect(await screen.findByText("Email is not valid")).toBeInTheDocument();
	});

	it("shows error message when confirm email field does not match email field", async () => {
		render(
			<QueryClientProvider client={queryClient}>
				<RequestInviteForm />
			</QueryClientProvider>
		);
		fireEvent.change(screen.getByPlaceholderText("Email"), {
			target: { value: "mock1@airwallex.com" },
		});
		fireEvent.change(screen.getByPlaceholderText("Confirm Email"), {
			target: { value: "mock2@airwallex.com" },
		});
		fireEvent.click(screen.getByTestId("submit-button"));
		expect(await screen.findByText("Emails must match")).toBeInTheDocument();
	});

	it("show the error message if validation passed but server returns 400", async () => {
		render(
			<QueryClientProvider client={queryClient}>
				<RequestInviteForm />
			</QueryClientProvider>
		);
		fireEvent.change(screen.getByPlaceholderText("Full Name"), {
			target: { value: "Stephen Shi" },
		});
		fireEvent.change(screen.getByPlaceholderText("Email"), {
			target: { value: "usedemail@airwallex.com" },
		});
		fireEvent.change(screen.getByPlaceholderText("Confirm Email"), {
			target: { value: "usedemail@airwallex.com" },
		});

		fireEvent.click(screen.getByTestId("submit-button"));

		await waitFor(
			() => {
				expect(
					screen.getByText("Request failed with status code 400")
				).toBeInTheDocument();
			},
			{ timeout: 3000 }
		);
	});
});
