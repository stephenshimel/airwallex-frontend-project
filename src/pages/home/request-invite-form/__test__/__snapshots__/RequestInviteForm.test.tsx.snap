// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RequestInviteForm initial status matches the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class="css-0"
      >
        <h2
          class="css-18dyf6a"
        >
          Request an invite
        </h2>
        <div
          class="css-108trlw"
          data-testid="separator-line"
        />
        <div
          class="css-2ycnf4"
        >
          <input
            class="css-1xshicx"
            name="name"
            placeholder="Full name"
          />
          <input
            class="css-1xshicx"
            name="email"
            placeholder="Email"
          />
          <input
            class="css-1xshicx"
            name="confirmEmail"
            placeholder="Confirm Email"
          />
        </div>
        <button
          class="css-yaz3o2"
          data-testid="submit-button"
          type="submit"
        >
          Send
        </button>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class="css-0"
    >
      <h2
        class="css-18dyf6a"
      >
        Request an invite
      </h2>
      <div
        class="css-108trlw"
        data-testid="separator-line"
      />
      <div
        class="css-2ycnf4"
      >
        <input
          class="css-1xshicx"
          name="name"
          placeholder="Full name"
        />
        <input
          class="css-1xshicx"
          name="email"
          placeholder="Email"
        />
        <input
          class="css-1xshicx"
          name="confirmEmail"
          placeholder="Confirm Email"
        />
      </div>
      <button
        class="css-yaz3o2"
        data-testid="submit-button"
        type="submit"
      >
        Send
      </button>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`RequestInviteForm show the success page, if the form is submitted, passed all validation rules, and server response is 200 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class="css-0"
      >
        <h2
          class="css-18dyf6a"
        >
          All done!
        </h2>
        <div
          class="css-108trlw"
          data-testid="separator-line"
        />
        <h3
          class="css-bao8lp"
        >
          You will be one of the first to experience Broccoli & Co. when w launch
        </h3>
        <button
          class="css-yaz3o2"
          data-testid="submit-button"
          type="submit"
        >
          OK
        </button>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class="css-0"
    >
      <h2
        class="css-18dyf6a"
      >
        All done!
      </h2>
      <div
        class="css-108trlw"
        data-testid="separator-line"
      />
      <h3
        class="css-bao8lp"
      >
        You will be one of the first to experience Broccoli & Co. when w launch
      </h3>
      <button
        class="css-yaz3o2"
        data-testid="submit-button"
        type="submit"
      >
        OK
      </button>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
